rewrite-xml/src/main/java/org/openrewrite/xml/trait/ContainsPackageReference.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.xml.trait;
+
+import lombok.Value;
+import org.jspecify.annotations.Nullable;
+import org.openrewrite.Cursor;
+import org.openrewrite.Tree;
+import org.openrewrite.trait.SimpleTraitMatcher;
+import org.openrewrite.trait.Trait;
+import org.openrewrite.xml.tree.Xml;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+
+@Value
+public class ContainsPackageReference implements Trait<Tree> {
+    Cursor cursor;
+
+    public @Nullable String getValue() {
+        if (getTree() instanceof Xml.Attribute) {
+            Xml.Attribute attribute = (Xml.Attribute) getTree();
+            return attribute.getValueAsString();
+        } else if (getTree() instanceof Xml.Tag) {
+            Xml.Tag tag = (Xml.Tag) getTree();
+            if (tag.getValue().isPresent()) {
+                return tag.getValue().get();
+            }
+        }
+        return null;
+    }
+
+
+    public static class Matcher extends SimpleTraitMatcher<ContainsPackageReference> {
+        private final Pattern PACKAGE_OR_TYPE_REFERENCE = Pattern.compile("^([a-zA-Z_][a-zA-Z0-9_]*)(\\.[a-zA-Z_][a-zA-Z0-9_]*)*(\\.[A-Z][a-zA-Z0-9_]*|\\$[A-Z][a-zA-Z0-9_]*)*(\\.\\*)?$");
+        private final List<String> ATTRIBUTES_THAT_REFERENCE_PACKAGE_OR_TYPE = Arrays.asList("class", "type");
```

Comment:
I think regular expressions are an unsafe way to go about it, but I'm not exactly sure what a better approach would be.  XPath would be a little more predictable, but also feel out of place here to have Spring/Micronaut/Quarkus/EE expressions all in here to be picked up when changing types.

Perhaps @knut has some suggestions on using service discovery where frameworks specific recipe modules can provide a mechanism to provide additional paths to check and convert.

Reply:
I think it would probably be best if we could register some kind of "extensions" with parsers like those for XML, YAML, Properties, etc. So we could then already while ingesting the source file attach some kind of `JavaTypeReference` marker to the LST elements. Then the `ChangeType` recipe could check if the `SourceFile` if it has some kind of `HasJavaTypeReferences` marker or possibly it even uses the `TypesInUse` we use for `JavaSourceFile`.

An additional complication may be that these "parser extensions" might have to be registered dynamically (e.g. depending on whether the project is a Spring project or not).

This could quickly end up being quite a bit of work. But I get the impression that this would long-term be the better and more sustainable solution. Maybe we can still do something like this, but start out small?